src/vendor/requests.js

```diff
@@ -0,0 +1,103 @@
+/* global fetch */
+import { parse } from 'query-string';
+import { leaves, toPairs } from './vectors';
+import { Log } from './logs';
+import { Data } from './Data';
+import { exists, isArray, isString, toArray } from './utils';
+import strings from './strings';
+
+function fromQueryString(query) {
+  const decode = v => {
+    if (isArray(v)) return v.map(decode);
+
+    if (v === null || v === '') return true;
+
+    try {
+      return JSON.parse(v);
+    } catch (e) {
+      return v;
+    }
+  };
+
+  return toPairs(parse(query))
+    .map(decode)
+    .fromLeaves();
+}
+
+function toQueryString(value) {
+  const e = vv => encodeURIComponent(vv).replace(/[%]20/g, '+');
+  const encode = (v, k) =>
+    toArray(v)
+      .filter(exists)
+      .map(vv => {
+        if (vv === true) return e(k);
+
+        if (isString(vv)) {
+          try {
+            JSON.parse(vv);
+
+            return `${e(k)}=${e(JSON.stringify(vv))}`;
+          } catch (e) {
+            // USE STANDARD ENCODING
+          }
+        }
+
+        return `${e(k)}=${e(vv)}`;
+      })
+      .join('&');
+  const temp = leaves;
+
+  return temp(value)
+    .map(encode)
+    .join('&');
+}
+
+const jsonHeaders = {
+  Accept: 'application/json',
+};
+const fetchJson = async url => {
+  const response = await fetch(url, jsonHeaders);
+
+  if (!response) return null;
+
+  if (response.status !== 200) {
+    Log.error('{{status}} when calling {{url}}', {
+      url,
+      status: response.status,
+    });
+  }
+
+  try {
+    return response.json();
+  } catch (error) {
+    Log.error(`Problem parsing ${response.text()}`);
+    // Log.error("Problem parsing {{text}}", {text: response.text()});
+  }
+};
+
+const URL = ({ path, query, fragment }) => {
```

Comment:
Thoughts on using https://developer.mozilla.org/en-US/docs/Web/API/URL instead?

Reply:
Excellent! Thank you very much for pointing this API out. It does a lot of what I would like to see in the future, like identifying protocol, I appreciate your effort to find a viable alternative.

`URL` has a few drawbacks:

1. `URL` seems to represent slices of the URL string, rather than a serialization-independent API.  For example, `URL.hash` includes the hash (`#`) symbol, which is redundant.  Same with `URL.search`, which includes the question mark (`?`).  `URL.protocol` includes the colon (`:`).
2. `URL` uses unconventional names: [The RFC](https://tools.ietf.org/html/rfc3986) uses "scheme", "query" and "fragment", while `URL` uses "protocol", "search", and "hash".
3. `URL` does not set `query` - One main purpose of my function is to accept a `query` object to build the query part of the URL. The `URL` you point to only allows reading such an object, plus I am not certain that the returned objects are "properly" parsed.
4. `URL` lacks parametric paths - Many RESTful APIs demand segments of the `path` consist of meaningful parameter values.  For example, the `repo` parameter is put in the 3rd step of a 5-step path: `api/project/${repo}/performance/signatures/`.  My class allows parameterization of the `path` by accepting an array of strings, some variable, which will be assembled to a proper path.